### PR TITLE
update linenoise

### PR DESCRIPTION
**Pull request by ringabout**
fetch the upstream updates

based on this commit 

https://github.com/antirez/linenoise/commit/97d2850af13c339369093b78abe5265845d78220

### CODE REVIEW

File: lib/wrappers/linenoise/linenoise.c [hunk 1]
```diff
@@ -600,7 +625,7 @@ static void refreshMultiLine(struct linenoiseState *l) {
     rpos2 = (plen+l->pos+l->cols)/l->cols; /* current cursor relative row. */
     lndebug("rpos2 %d", rpos2);
 
-    /* Go up till we reach the expected position. */
+    /* Go up till we reach the expected positon. */
```
Comment (ringabout) [inline lib/wrappers/linenoise/linenoise.c]:
```suggestion
    /* Go up till we reach the expected position. */
```

File: lib/wrappers/linenoise/linenoise.c [hunk 2]
```diff
@@ -485,6 +499,8 @@ void refreshShowHints(struct abuf *ab, struct linenoiseState *l, int plen) {
             if (bold == 1 && color == -1) color = 37;
             if (color != -1 || bold != 0)
                 snprintf(seq,64,"\033[%d;%d;49m",bold,color);
+            else
+                seq[0] = '\0';
```
Comment (ringabout) [inline lib/wrappers/linenoise/linenoise.c:503]: It seems to have a small fix here

**Comment by ringabout** [comment]
It is also helpful for implementing `readPasswordFromStdin` if needed.

**Comment by ringabout** [comment]
Yeah, is this module rarely used?

**Comment by Araq** [comment]
Please rewrite all the C code in Nim. We don't need more "static" variables, it's a desaster for multi-threading.

**Comment by timotheecour** [comment]
* does this fix https://github.com/nim-lang/Nim/issues/16449 ? (not seeing this issue mentioned here)
* as mentioned in https://github.com/nim-lang/RFCs/issues/206#issuecomment-717915125, linenoise should be either a submodule or equivalent, it's a bad idea to attempt to merge 2 unrelated git histories; the correct way to do this is to write a fork of https://github.com/antirez/linenoise that's kept up to date with upstream and that we clone inside nim during (for eg) `koch boot`. Anything other than that means updating from upstream and keeping local changes is hard.

> Please rewrite all the C code in Nim

That said, I kind of agree with the need for rewrite, although it's a bit unfair to dump it all on this PR's author. Long term this is absolutely the way to go though, for several reasons:

* https://github.com/antirez/linenoise is stable, not getting much updates/new features
* it's still very small
* it has multi-threading issues as noted
* a rewrite in nim would be much smaller than the already small C code, and would allow us to add more features
* there are no dependencies other than standard ones

I had actually done exactly that in D some time ago, and it was not hard, and it did allow implementing more features easily in a more high-level langauge.

In nim, this could also be done in multiple stages instead of a one-time conversion, thanks to `{.emit.}`.

> Yeah, is this module rarely used?

no, it's useful anytime you need interactive CLI input in a program (eg for a repl, eg inim)

**Comment by ringabout** [comment]
I could port them when I have free time.

I will also consider porting a better and modern one:
https://github.com/AmokHuginnsson/replxx

**Comment by ringabout** [comment]
It seems that there are already two Pure Nim version of linenoise
https://github.com/jangko/nim-noise
https://github.com/de-odex/linenoise-nim

**Comment by timotheecour** [comment]
@araq @xflywind instead of rewriting something that already exists in nimble (in multiple packages in fact), how about this plan:
* implement https://github.com/nim-lang/RFCs/issues/291 (--path:prefix:path)
* users can then override std/linenoise with their favorite drop-in replacement, eg:
```nim
nimble install noise # suppose it has a `linenoisecompat` module with same interface as std/linenoise
nim --path:std/linenoise:pkg/noise/linenoisecompat main
```

* `koch boot` can itself do: `nimble install noise` and we can add `--path:std/linenoise:pkg/noise/linenoisecompat` to nim/config/config.nims, which is overridable in user config

then it all works, and even `nim secret` can keep working with linenoise without users having to change anything; but users can override linenoise with another nimble package that offers a compatibility module, eg nim-noise, linenoise-nim etc.

* we can then deprecate `lib/wrappers/linenoise/linenoise.nim` and after a deprecation period remove `lib/wrappers/linenoise/*`

## note
this approach generalizes to other use cases (there's nothing specific about linenoise) where we're better off reusing nimble packages than re-inventing (poorly) the wheel in stdlib, thus avoiding duplicating effort

**Pull request closed without merging**